packages/node_modules/pouchdb-errors/src/index.js

```diff
@@ -44,26 +44,13 @@ var MISSING_STUB = new PouchError(412, 'missing_stub', 'A pre-existing attachmen
 var INVALID_URL = new PouchError(413, 'invalid_url', 'Provided URL is invalid');
 
 function createError(error, reason) {
-  function CustomPouchError(reason) {
-    // inherit error properties from our parent error manually
-    // so as to allow proper JSON parsing.
-    var names = Object.getOwnPropertyNames(error);
-    for (var i = 0, len = names.length; i < len; i++) {
-      if (typeof error[names[i]] !== 'function') {
-        this[names[i]] = error[names[i]];
-      }
-    }
-
-    if (this.stack === undefined) {
-      this.stack = (new Error()).stack;
-    }
-
-    if (reason !== undefined) {
-      this.reason = reason;
-    }
+  // inherit error properties from our parent error manually
+  // so as to allow proper JSON parsing.
+  var err = new PouchError(error.status, error.name, error.message);
+  if (reason !== undefined) {
+    err.reason = reason;
```

Comment:
Does this change risk losing non-standard props of the reanimated error message when `createError()` is called on something other than one of the constants defined in this file?  E.g.:

```
packages/node_modules/pouchdb-find/src/adapters/http/index.js:16:    const pouchError = createError(json);
packages/node_modules/pouchdb-replication/src/replicate.js:299:      fatalError = createError(fatalError);
```

Reply:
Ah ok, I thought it was ok to push as all tests passed.
Yeah, non-standard props will be lost by this change.
So, we should copy props except for `stack`.

Reply:
It doesn't look like there's much testing around `PouchError` or `createError()`, so maybe it's not surprising that tests were passing.  There's also some arcane stuff happening in both of those locations, so it would be good to get more tests in place before making changes.  Tests for the property copying is one example.

That said, this PR looks like an improvement and probably it's ok to break behaviour which is weird, undocumented, and probably the result of legacy javsascript distorted by years of refactoring...

Some other examples of behaviour that might be worth testing:

* it looks like `CustomPouchError.prototype = PouchError.prototype` propagates `PouchError.toString` to the result of `createError`, but this behaviour isn't tested
* it looks like an instance of `PouchError` will have property `.error` set to `true`, but a `CustomPouchError` will not, unless it was created from a `PouchError`.  Maybe `createError` is only meant to be called on instances of `PouchError`?  Either way, this PR currently changes that behaviour.

Maybe `CustomPouchError` should be a publicly-accessible class?  Ref https://github.com/pouchdb/pouchdb/issues/1167#issuecomment-1426502663

etc.

Reply:
I think it'd be better to export `PouchError` instead of `CustomPouchError`.

Yeah, it looks like `createError` is basically meant to be called on instances of `PouchError` as far as I can see:

```js
❯ rg createError
pouchdb-adapter-utils/src/updateDoc.js
1:import { createError, REV_CONFLICT } from 'pouchdb-errors';
41:    var err = createError(REV_CONFLICT);

pouchdb-adapter-utils/src/preprocessAttachments.js
8:import { createError, BAD_ARG } from 'pouchdb-errors';
15:    var err = createError(BAD_ARG,

pouchdb-adapter-utils/src/parseDoc.js
6:  createError
52:    return createError(INVALID_REV);
148:        var error = createError(DOC_VALIDATION, key);

pouchdb-changes-filter/src/index.js
4:  createError,
54:      var err = createError(BAD_REQUEST,
72:        return callback(createError(MISSING_DOC,
98:        return callback(createError(MISSING_DOC,

pouchdb-adapter-utils/src/processDocs.js
1:import { createError, MISSING_DOC, REV_CONFLICT } from 'pouchdb-errors';
25:      results[resultsIdx] = createError(MISSING_DOC, 'deleted');
33:      var err = createError(REV_CONFLICT);

pouchdb-replication/src/replicate.js
6:import { createError } from 'pouchdb-errors';
299:      fatalError = createError(fatalError);

pouchdb-utils/src/invalidIdError.js
5:  createError
16:    err = createError(MISSING_ID);
18:    err = createError(INVALID_ID);
20:    err = createError(RESERVED_ID);

pouchdb-replication/src/replicateWrapper.js
4:import { createError, BAD_REQUEST } from 'pouchdb-errors';
26:    throw createError(BAD_REQUEST,

pouchdb-adapter-leveldb-core/src/index.js
54:  createError
374:        return callback(createError(MISSING_DOC, 'missing'));
382:          return callback(createError(MISSING_DOC, "deleted"));
392:          return callback(createError(MISSING_DOC));
476:          var err = createError(MISSING_STUB,
656:            callback(createError(BAD_ARG,
1167:      return callback(createError(NOT_OPEN));
1195:        callback(createError(MISSING_DOC));
1363:        callback(createError(MISSING_DOC));
1396:        return callback(createError(REV_CONFLICT));
1399:        return callback(createError(REV_CONFLICT));
1454:          return callback(createError(MISSING_DOC));
1458:        return callback(createError(REV_CONFLICT));

pouchdb-core/src/adapter.js
36:  createError
210:        return callback(createError(NOT_AN_OBJECT));
221:        return cb(createError(NOT_AN_OBJECT));
225:        return cb(createError(INVALID_REV));
301:          throw createError(REV_CONFLICT);
324:          callback(createError(REV_CONFLICT));
505:        return cb(createError(INVALID_ID));
571:                return cb(createError(INVALID_REV));
576:            return cb(createError(UNKNOWN_ERROR, 'function_clause'));
708:          return callback(createError(MISSING_DOC));
734:          callback(createError(QUERY_PARSE_ERROR,
789:        return callback(createError(MISSING_BULK_DOCS));
795:          return callback(createError(NOT_AN_OBJECT));
798:          return callback(createError(INVALID_REV));
815:        return callback(createError(BAD_REQUEST, attachmentError));
1008:    return callback(createError(UNKNOWN_ERROR, 'Purge is not implemented in the ' + this.adapter + ' adapter.'));
1017:      return callback(createError(MISSING_DOC));

pouchdb-find/src/adapters/http/index.js
1:import { createError, generateErrorFromResponse } from 'pouchdb-errors';
16:    const pouchError = createError(json);

pouchdb-adapter-idb/src/allDocs.js
1:import { createError, IDB_ERROR } from 'pouchdb-errors';
86:      return callback(createError(IDB_ERROR,

pouchdb-errors/src/index.js
46:function createError(error, reason) {
120:  createError,

pouchdb-adapter-idb/src/utils.js
2:import { createError, IDB_ERROR } from 'pouchdb-errors';
24:    callback(createError(IDB_ERROR, message, evt.type));

pouchdb-adapter-http/src/index.js
8:  createError,
696:        return callback(createError(BAD_ARG,

pouchdb-adapter-indexeddb/src/getRevisionTree.js
3:import { createError, MISSING_DOC } from 'pouchdb-errors';
15:      callback(createError(MISSING_DOC));

pouchdb-for-coverage/src/utils.js
43:  createError,
59:  createError,

pouchdb-adapter-indexeddb/src/allDocs.js
3:import { createError, IDB_ERROR } from 'pouchdb-errors';
57:  callback(createError(IDB_ERROR, err.name, err.message));

pouchdb-adapter-idb/src/index.js
24:  createError
346:        err = createError(MISSING_DOC, 'missing');
355:          err = createError(MISSING_DOC, "deleted");
371:          err = createError(MISSING_DOC, 'missing');
455:        callback(createError(MISSING_DOC));
514:        callback(createError(MISSING_DOC));
559:          callback(createError(REV_CONFLICT));
574:        callback(createError(REV_CONFLICT));
614:        callback(createError(MISSING_DOC));
813:    callback(createError(IDB_ERROR, msg));

pouchdb-adapter-idb/src/bulkDocs.js
1:import { createError, MISSING_STUB } from 'pouchdb-errors';
176:        var err = createError(MISSING_STUB,

pouchdb-adapter-indexeddb/src/util.js
3:import { createError, IDB_ERROR } from 'pouchdb-errors';
20:    callback(createError(IDB_ERROR, message, evt.type));

pouchdb-adapter-indexeddb/src/get.js
3:import { createError, MISSING_DOC } from 'pouchdb-errors';
24:      callback(createError(MISSING_DOC, 'missing'));

pouchdb-adapter-indexeddb/src/bulkDocs.js
4:  createError,
85:        newDoc = createError(MISSING_DOC, 'deleted');
91:        newDoc = createError(REV_CONFLICT);
176:      return createError(REV_CONFLICT);
257:            error = createError(MISSING_STUB);
325:        return Promise.reject(createError(BAD_ARG, 'Attachment is not a valid base64 string'));
410:        callback(error || createError(UNKNOWN_ERROR, 'transaction was aborted'));

pouchdb-utils/src/filterChange.js
1:import { createError, BAD_REQUEST } from 'pouchdb-errors';
8:    return createError(BAD_REQUEST, msg);

pouchdb-adapter-indexeddb/src/getLocal.js
1:import { createError, MISSING_DOC } from 'pouchdb-errors';
17:      callback(createError(MISSING_DOC, 'missing'));

pouchdb-adapter-indexeddb/src/getAttachment.js
4:import { createError, MISSING_DOC } from 'pouchdb-errors';
13:    cb(createError(MISSING_DOC, 'missing'));
```

except for:

```js
pouchdb-replication/src/replicate.js
6:import { createError } from 'pouchdb-errors';
299:      fatalError = createError(fatalError);

pouchdb-find/src/adapters/http/index.js
1:import { createError, generateErrorFromResponse } from 'pouchdb-errors';
16:    const pouchError = createError(json);
```

So, maybe copying non-standard props is only for these two use cases? 🤔 I'll look into how PouchDB handles these non-standard error props.

Reply:
Another question is: are the `const` error instances of `PouchError` ever used _without_ passing them to `createError()` first?  If not, we can probably simplify everything a lot.

Reply:
I guess `createError` is called for getting a correct stack trace thus using the error constants directly is wrong here.
Sounds nice. Can you give me an idea for the simplification?

Reply:
Something like this: https://github.com/pouchdb/pouchdb/pull/9028/files